File: tests/Python/test_validation.py
import numpy
import pytest

abcpp = pytest.importorskip("abcpp")


def test_dimension_mismatch_raises(toy_data):
    with pytest.raises(Exception, match="target summary dimension"):
        abcpp.abc(
            target=toy_data["target"][:-1],
            params=toy_data["param"],
            sumstats=toy_data["sumstat"],
            control={"tol": 0.10},
        )

    with pytest.raises(Exception, match="param.rows"):
        abcpp.abc(
            target=toy_data["target"],
            params=toy_data["param"][:-1, :],
            sumstats=toy_data["sumstat"],
            control={"tol": 0.10},
        )


def test_invalid_method_raises(toy_data):
    with pytest.raises(Exception, match="Unknown ABC method"):
        abcpp.abc(
            target=toy_data["target"],
            params=toy_data["param"],
            sumstats=toy_data["sumstat"],
            control={"method": "bad_method", "tol": 0.10},
        )


def test_invalid_reduction_raises(toy_data):
    with pytest.raises(Exception, match="Unknown summary reduction"):
        abcpp.abc(
            target=toy_data["target"],
            params=toy_data["param"],
            sumstats=toy_data["sumstat"],
            control={"tol": 0.10, "reduction": "bad_reduction"},
        )


def test_invalid_tolerance_raises(toy_data):
    with pytest.raises(Exception, match="tol must be"):
        abcpp.abc(
            target=toy_data["target"],
            params=toy_data["param"],
            sumstats=toy_data["sumstat"],
            control={"tol": 0},
        )


def test_matrix_target_and_stacked_summary_statistics_work(toy_data):
    target_matrix = numpy.asarray([toy_data["target"][:2]])
    sumstat_matrix = toy_data["sumstat"][:, :2]

    result_none = abcpp.abc(
        target=target_matrix,
        params=toy_data["param"],
        sumstats=sumstat_matrix,
        control={"tol": 0.10, "reduction": "none"},
    )

    assert result_none["numstat"] == 2

    stacked_sumstat = numpy.reshape(sumstat_matrix, (-1, 1))
    target_stacked = numpy.reshape(toy_data["target"][:2], (2, 1))

    result_pls = abcpp.abc(
        target=target_stacked,
        params=toy_data["param"],
        sumstats=stacked_sumstat,
        control={
            "method": "loclinear",
            "tol": 0.30,
            "hcorr": False,
            "reduction": "pls",
            "n_comp": 1,
        },
    )

    assert result_pls["numstat"] == 1
    assert result_pls["reduction"]["method"] == "PLS"


def test_matrix_target_and_list_summary_statistics_work():
    param = numpy.arange(5, dtype=float).reshape((-1, 1))
    sumstats = [
        numpy.asarray(
            [[value, value + 0.1], [value + 0.2, value + 0.3]],
            dtype=float,
        )
        for value in range(5)
    ]
    target = numpy.asarray([[2.0, 2.1], [2.2, 2.3]], dtype=float)

    result = abcpp.abc(
        target=target,
        params=param,
        sumstats=sumstats,
        control={"tol": 0.40, "reduction": "none"},
    )

    assert result["numstat"] == 4
    assert result["unadj_values"].shape[0] == 2
    assert result["ss"].shape[1] == 4
    assert len(abcpp.summary(result)["columns"]) == 1


def test_matrix_target_and_dict_summary_statistics_work():
    param = numpy.arange(5, dtype=float).reshape((-1, 1))
    sumstats = {
        f"sim_{value}": numpy.asarray(
            [[value, value + 0.1], [value + 0.2, value + 0.3]],
            dtype=float,
        )
        for value in range(5)
    }
    target = numpy.asarray([[2.0, 2.1], [2.2, 2.3]], dtype=float)

    result = abcpp.abc(
        target=target,
        params=param,
        sumstats=sumstats,
        control={"tol": 0.40, "reduction": "none"},
    )

    assert result["numstat"] == 4
    assert result["unadj_values"].shape[0] == 2
    assert result["ss"].shape[1] == 4


def test_one_row_parameter_matrix_is_treated_as_one_parameter(toy_data):
    param_row = numpy.reshape(toy_data["param"][:, 0], (1, -1))

    result = abcpp.abc(
        target=toy_data["target"],
        params=param_row,
        sumstats=toy_data["sumstat"],
        control={"tol": 0.10},
    )

    assert result["numparam"] == 1
    assert result["unadj_values"].shape[1] == 1


def test_one_dimensional_param_and_sumstat_are_reshaped():
    param = numpy.linspace(0.0, 1.0, 30)
    sumstat = param + 0.01

    result = abcpp.abc(
        target=[0.51],
        params=param,
        sumstats=sumstat,
        control={"tol": 0.20},
    )

    assert result["numparam"] == 1
    assert result["numstat"] == 1
    assert result["unadj_values"].shape[1] == 1

File: tests/R/testthat/test-validation.R
testthat::test_that("dimension mismatch raises an error", {
  data <- abcpp_test_data()

  testthat::expect_error(
    abcpp::abc(
      target = data$target[-1L],
      param = data$param,
      sumstat = data$sumstat,
      tol = 0.10,
      method = "rejection"
    ),
    "target summary dimension"
  )

  testthat::expect_error(
    abcpp::abc(
      target = data$target,
      param = data$param[-1L, ],
      sumstat = data$sumstat,
      tol = 0.10,
      method = "rejection"
    ),
    "param.rows"
  )
})

testthat::test_that("invalid method and reduction raise errors", {
  data <- abcpp_test_data()

  testthat::expect_error(
    abcpp::abc(
      target = data$target,
      param = data$param,
      sumstat = data$sumstat,
      tol = 0.10,
      method = "bad_method"
    ),
    "Unknown ABC method"
  )

  testthat::expect_error(
    abcpp::abc(
      target = data$target,
      param = data$param,
      sumstat = data$sumstat,
      tol = 0.10,
      method = "rejection",
      reduce = "bad_reduction"
    ),
    "Unknown summary reduction"
  )
})

testthat::test_that("invalid tolerance raises an error", {
  data <- abcpp_test_data()

  testthat::expect_error(
    abcpp::abc(
      target = data$target,
      param = data$param,
      sumstat = data$sumstat,
      tol = 0,
      method = "rejection"
    ),
    "tol must be"
  )
})

testthat::test_that("subset changes accepted rows consistently", {
  data <- abcpp_test_data()
  subset <- base::rep(TRUE, base::nrow(data$param))
  subset[base::seq(from = 1L, to = base::length(subset), by = 2L)] <- FALSE

  result <- abcpp::abc(
    target = data$target,
    param = data$param,
    sumstat = data$sumstat,
    tol = 0.10,
    method = "rejection",
    transf = base::rep("none", 2L),
    subset = subset
  )

  testthat::expect_true(base::all(!result$region[!subset]))
  testthat::expect_true(base::all(result$na.action == subset))
})

testthat::test_that("matrix target and stacked summary statistics work", {
  data <- abcpp_test_data()
  target_matrix <- base::matrix(data$target[1L:2L], nrow = 1L)
  sumstat_matrix <- data$sumstat[, 1L:2L]

  result_none <- abcpp::abc(
    target = target_matrix,
    param = data$param,
    sumstat = sumstat_matrix,
    tol = 0.10,
    method = "rejection",
    reduction = "none"
  )

  testthat::expect_equal(result_none$numstat, 2L)

  stacked_sumstat <- base::matrix(
    base::as.vector(base::t(sumstat_matrix)),
    ncol = 1L
  )
  target_stacked <- base::matrix(data$target[1L:2L], nrow = 2L)

  result_pls <- abcpp::abc(
    target = target_stacked,
    param = data$param,
    sumstat = stacked_sumstat,
    tol = 0.30,
    method = "loclinear",
    hcorr = FALSE,
    reduction = "pls",
    ncomp = 1L
  )

  testthat::expect_equal(result_pls$numstat, 1L)
  testthat::expect_identical(result_pls$reduction$method, "PLS")
})

testthat::test_that("matrix target and list summary statistics work", {
  param <- base::matrix(seq(0, 4), ncol = 1L)
  sumstats <- base::lapply(seq(0, 4), function(value) {
    base::matrix(
      base::c(value, value + 0.1, value + 0.2, value + 0.3),
      nrow = 2L,
      byrow = TRUE
    )
  })
  target <- base::matrix(
    base::c(2, 2.1, 2.2, 2.3),
    nrow = 2L,
    byrow = TRUE
  )

  result <- abcpp::abc(
    target = target,
    param = param,
    sumstat = sumstats,
    tol = 0.40,
    method = "rejection",
    reduction = "none"
  )

  testthat::expect_equal(result$numstat, 4L)
  testthat::expect_equal(base::nrow(result$unadj.values), 2L)
  testthat::expect_equal(base::ncol(result$ss), 4L)
  testthat::expect_s3_class(base::summary(result), "summary.abcpp")
})

testthat::test_that("one-row parameter matrix is treated as one parameter", {
  data <- abcpp_test_data()
  param_row <- base::matrix(data$param[, 1L], nrow = 1L)

  result <- abcpp::abc(
    target = data$target,
    param = param_row,
    sumstat = data$sumstat,
    tol = 0.10,
    method = "rejection"
  )

  testthat::expect_equal(result$numparam, 1L)
  testthat::expect_equal(base::ncol(result$unadj.values), 1L)
})
